src/azure-cli-core/azure/cli/core/cloud.py

```diff
@@ -0,0 +1,181 @@
+#---------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+#---------------------------------------------------------------------------------------------
+
+import os
+from six.moves import configparser
+
+from azure.cli.core._config import GLOBAL_CONFIG_DIR
+
+CUSTOM_CLOUD_CONFIG_FILE = os.path.join(GLOBAL_CONFIG_DIR, 'clouds.config')
+
+class CloudNotRegisteredException(Exception):
+    def __init__(self, cloud_name):
+        super(CloudNotRegisteredException, self).__init__(cloud_name)
+        self.cloud_name = cloud_name
+    def __str__(self):
+        return "The cloud '{}' is not registered.".format(self.cloud_name)
+
+class CloudAlreadyRegisteredException(Exception):
+    def __init__(self, cloud_name):
+        super(CloudAlreadyRegisteredException, self).__init__(cloud_name)
+        self.cloud_name = cloud_name
+    def __str__(self):
+        return "The cloud '{}' is already registered.".format(self.cloud_name)
+
+class CannotUnregisterCloudException(Exception):
+    pass
+
+class CloudEndpointUrl(object): # pylint: disable=too-few-public-methods
+    MANAGEMENT = 'management'
+    RESOURCE_MANAGER = 'resource_manager'
+    SQL_MANAGEMENT = 'sql_management'
+    GALLERY = 'gallery'
+    ACTIVE_DIRECTORY = 'active_directory'
+    ACTIVE_DIRECTORY_RESOURCE_ID = 'active_directory_resource_id'
+    ACTIVE_DIRECTORY_GRAPH_RESOURCE_ID = 'active_directory_graph_resource_id'
+
+class CloudParam(object): # pylint: disable=too-few-public-methods
+    SQL_SERVER_HOSTNAME_SUFFIX = 'sql_server_hostname_suffix'
+    STORAGE_ENDPOINT_SUFFIX = 'storage_endpoint_suffix'
+    KEYVAULT_DNS_SUFFIX = 'keyvault_dns_suffix'
+    AZURE_DATALAKE_STORE_FILE_SYSTEM_ENDPOINT_SUFFIX = 'azure_datalake_store_file_system_endpoint_suffix' #pylint: disable=line-too-long
+    AZURE_DATALAKE_ANALYTICS_CATALOG_AND_JOB_ENDPOINT_SUFFIX = 'azure_datalake_analytics_catalog_and_job_endpoint_suffix' #pylint: disable=line-too-long
+
+class Cloud(object): # pylint: disable=too-few-public-methods
+    """ Represents an Azure Cloud instance """
+
+    def __init__(self, name, endpoints=None, params=None):
+        self.name = name
+        self.endpoints = endpoints
+        self.params = params
+
+AZURE_PUBLIC_CLOUD = Cloud('AzureCloud', endpoints={
```

Comment:
The configuration for the 'common clouds' will not change and shouldn't be user editable.
Saving it to a file would allow users to edit it and give the impression that it's safe to do so.

Also, if we wanted to save it to a file on install, the code to save it to file would have to be somewhere... We can't just include the in the CLI install as it would get installed in `python-X/site-packages/azure/cli/core/cloud-something` instead of `~/.azure/cloud.config` where the custom cloud information goes.